contributions/catalog/controllers/contribute.py

```diff
@@ -212,21 +219,135 @@ def contribution_edit(contribution_id):
+        is_reviewer = False
         username = session["username"]
         headers = requestutil.get_header_using_session(session)
-        is_editable = adminutil.check_if_reviewer(username, headers)
+
+        # check if the user is in contribution's admins
+        if username in the_json_res["contributionAdmins"]:
+            is_editable = True
+
+        # check if the user is a reviewer
+        is_reviewer = adminutil.check_if_reviewer(username, headers)
+
+        if is_reviewer:
+            is_editable = True
 
         # get capability list to create required capability list
         required_capability_list = requestutil.request_required_capability_list(headers)
 
         if is_editable:
-            return render_template('contribute/contribute.html', required_capabilities=required_capability_list,
+            return render_template('contribute/contribute.html', is_reviewer=is_reviewer, required_capabilities=required_capability_list,
                                    is_editable=is_editable, user=session["name"],
                                    token=session['oauth_token']['access_token'], post=the_json_res)
         else:
             s = "You don't have a permission to edit the contribution."
             return render_template('contribute/error.html', error_msg=s)
 
+@bp.route('/contributions/<contribution_id>/review', methods=['GET', 'POST'])
+@login_required
+def contribution_review(contribution_id):
+    username = session["username"]
+    headers = requestutil.get_header_using_session(session)
+
+    if request.method == 'POST':
+        is_put = False
+        contribution_id = None
+        result = request.form.to_dict(flat=False)
+
+        # check if it is PUT
+        try:
+            contribution_status = result["contribution_status"][0]
+            contribution_comment = result["contribution_reviewer_comment"][0]
+            contribution_id = result["contribution_id"][0]
+            is_put = True
+        except:
+            s = "There is a error in edit. The method is not an edit."
+            return render_template('contribute/error.html', error_msg=s)
+
+        if is_put:
+            # check reviewer id from token
+            reviewer_id = adminutil.get_reviewer_id(headers, username)
+            if reviewer_id.lower == "none":
+                s = "You don't have a permission to review the contribution."
+                return render_template('contribute/error.html', error_msg=s)
+            contribution_result = requestutil.request_single_contribution(headers, contribution_id)
+            contribution = json.loads(contribution_result.text)
+
+            # remove id from json_data
+            del contribution["id"]
+
+            currenttime = datetime.datetime.now()
+            currenttime = currenttime.strftime("%Y/%m/%dT%H:%M:%S")
+
+            # iterate reviews to find out the correct review location
+            review_loc = 0
+            if "review" in contribution.keys():
```

Comment:
This is also from local changes.
```suggestion
            if "review" in contribution.keys() and contribution["review"] is not None:
```
I think this check was needed to address another case when the review field is not existing in the contribution. Please check and apply as you consider appropriate. Thanks.

Reply:
added